### ISSUE TITLE
GeekBench reports unsolved symbol nvmlInit_v2 or fails to find "cuda" library on Linux

### ISSUE THREAD

**Issue by xzhangxa** [vosen/ZLUDA#19]
Hello,

I'm trying to run ZLUDA with the GeekBench on Ubuntu 20.04 as the README shows. The library is built successfully as libzluda.so is generated, using latest Intel compute runtime [20.47.18513](https://github.com/intel/compute-runtime/releases/tag/20.47.18513). When renaming libzluda.so to libcuda.so and set LD_LIBRARY_PATH to the path and run GeekBench geekbench_x86_64 -h, the following error shows along with other geekbench help messages:

```
LD_LIBRARY_PATH=~/projects/ZLUDA/target/debug ./geekbench_x86_64 -h                                                                               
...
[1202/133024:FATAL:src/halogen/cuda/cuda_library.cpp(1713)] Failed to load nvmlInit_v2: ./geekbench_x86_64: undefined symbol: nvmlInit_v2
[1202/133024:FATAL:src/halogen/cuda/cuda_library.cpp(1713)] Failed to load nvmlInit_v2: ./geekbench_x86_64: undefined symbol: nvmlInit_v2
...
```

Googled this symbol is from NVML libnvidia-ml.so, so I installed cuda toolkit 11.1 altogether to the system, but without attach a NV GPU and tried again. This time GeekBench reports it cannot find CUDA library:

```
[1202/131130:WARNING:src/halogen/cuda/cuda_library.cpp(1465)] Cannot find or load CUDA library.
```

Though the fake libcuda.so is still in LD_LIBRARY_PATH. If not rename libzluda.so to libcuda.so, it seems GeekBench finds the real libcuda.so from system path and reports cuInit return 100 (device not find), as expected.

Could you point out what I'm doing wrong to reproduce the GeekBench testing? Is it right to rename generated libzluda.so to libcuda.so, and is the cuda toolkit still needed even if there's no NV GPU on board? Thanks!

**Comment by vosen** [comment]
This is extremely strange, I tried a bunch of things on my machine and I can't reproduce it.
_The problem might be lack loading of CUDA `libcuda.so.1`_. I noticed that CUDA Runtime loads both `libcuda.so` and `libcuda.so.1`. But everything works fine on my machine if both ZLUDA `libcuda.so` and CUDA `libcuda.so.1` gets loaded (I have not observed `libcuda.so.1` being used, just loaded). Maybe you have some env variables or other settings that make CUDA Runtime actually use `libcuda.so.1`?

For the record, my setup is to have a directory named `debug_run` with symlinks with proper names to avoid constant renaming after builds, like so:
```
vosen@vosen-desktop:/media/vosen/01D3D293C1C33E60/dev/ZLUDA/target/debug_run$ ll
total 6
drwxrwxrwx 1 vosen vosen  368 gru  6 12:47 ./
drwxrwxrwx 1 vosen vosen 4096 lis 30 19:25 ../
lrwxrwxrwx 1 vosen vosen   20 lis 26 23:56 libcuda.so -> ../debug/libzluda.so*
lrwxrwxrwx 1 vosen vosen   20 lis 26 23:56 libcuda.so.1 -> ../debug/libzluda.so*
lrwxrwxrwx 1 vosen vosen   44 lis 20 18:49 libigc.so.1 -> /home/vosen/igc-tree/build/Debug/libigc.so.1*
```
`libigc.so.1` is there to debug IGC issues, should make no difference.

My full command to run GeekBench is:
```
LD_LIBRARY_PATH=/media/vosen/01D3D293C1C33E60/dev/ZLUDA/target/debug_run/ ./geekbench_x86_64 --no-upload --compute CUDA
```
You should skip `--no-upload` if you don't have a commercial version.

I can recommend you to try two things to debug the issue:
* See what is the result of CUDA calls with ltrace, like so:
  ```
   LD_LIBRARY_PATH=/media/vosen/01D3D293C1C33E60/dev/notCUDA/target/debug_run/ ltrace -x 'cu*@libcuda*' -L  ./geekbench_x86_64 -h
    ```
  Correct output should look like this (well, maybe without INTEL-MESA warning):
    ```
    Geekbench 5.2.3 : https://www.geekbench.com/

    Usage:

    ./geekbench_x86_64 [ options ]

    Options:

    -h, --help                print this message
    --unlock EMAIL KEY        unlock Geekbench using EMAIL and KEY
    --load FILE               load and display Geekbench result from FILE
    --save FILE               save Geekbench result to FILE
    --export-csv FILE         export Geekbench result as CSV to FILE
    --export-html FILE        export Geekbench result as HTML to FILE
    --export-json FILE        export Geekbench result as JSON to FILE
    --export-xml FILE         export Geekbench result as XML to FILE
    --export-text FILE        export Geekbench result as text to FILE
    --upload                  upload results to the Geekbench Browser
    --no-upload               do not upload results to the Geekbench Browser
    --cpu                     run the CPU benchmark
    --sysinfo                 display system information and exit
    cuInit@libcuda.so(0, 1, 1, 0)                                                                                                                     = 0
    cuDriverGetVersion@libcuda.so(0x7ffea589f79c, 0, 0, 0)                                                                                            = 0
    cuDeviceGetCount@libcuda.so(0x7ffea589f774, 0x459a190, 416, 384)                                                                                  = 0
    cuDeviceGetCount@libcuda.so(0x7ffea589f700, 0x7f9a, 0x7f9a, 0)                                                                                    = 0
    cuDeviceGet@libcuda.so(0x7ffea589f704, 0, 0, 0)                                                                                                   = 0
    cuDeviceGetAttribute@libcuda.so(0x7ffea589f6fc, 75, 0, 0)                                                                                         = 0
    INTEL-MESA: warning: Performance support disabled, consider sysctl dev.i915.perf_stream_paranoid=0


    --compute [API]           run the Compute benchmark
                                API can be one of: CUDA, OpenCL (default), 
    Vulkan
    --compute-list            list available compute platforms, devices and exit
    --compute-platform-id ID  use compute platform ID (default is 0)
    --compute-device-id ID    use compute device ID (default is 0)

    If no options are given, the default action is to run the CPU benchmark.

    +++ exited (status 0) +++
    ```

* Debug dynamic linker with `LD_DEBUG`:
  ```
  LD_LIBRARY_PATH=/media/vosen/01D3D293C1C33E60/dev/notCUDA/target/debug_run/ LD_DEBUG=libs  ~/Geekbench-5.2.3-Linux/./geekbench_x86_64 -h 2> ld_debug.log 
  ```
  Then inspect resulting `ld_debug.log` to see where from was `libcuda.so`/`libcuda.so.1` loaded

**Comment by xzhangxa** [comment]
@vosen Thanks for the help!

I figured out what's going on: the GeekBench binary does need `libnvidia-ml.so` for the symbol, but if I installed Debian testing package libnvidia-ml-dev it provides `/lib/x86_64-linux-gnu/libnvidia-ml.so.1` only. Previously I installed CUDA from Nvidia's apt source, maybe it also just provides `libnvidia-ml.so.1` but not `libnvidia-ml.so`. After making the symbol link it works now, the result is https://browser.geekbench.com/v5/compute/2011319.

So nothing wrong on ZLUDA side, it's just GeekBench executable needs libnvidia-ml.so so user still need to install it too. Closing this issue now.
